tidy3d/components/data/data_array.py

```diff
@@ -81,7 +81,10 @@ def __init__(self, data, *args, **kwargs):
 
     @property
     def tracers(self) -> Box:
-        if AUTOGRAD_KEY not in self.attrs and not isbox(self.data.flat[0]):  # no tracers
+        if self.data.size == 0:
+            return None
+        elif AUTOGRAD_KEY not in self.attrs and not isbox(self.data.flat[0]):
```

Comment:
might be slightly faster to assign `self.data.flat[0]` to a variable to avoid doing this twice. but I dont know how expensive this operation really is.

Reply:
This should be very fast because it's just a view into the object I believe? But actually yeah this might be a good case for the [walrus operator](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions). Was avoiding it because I thought it was introduced way after 3.8... time flies.

Reply:
not sure if we need to involve the walrus if we can just assign a variable?

Reply:
This seems pretty clean to me?
```python
@property
def tracers(self) -> Box:
    if self.data.size == 0:
        return None
    elif AUTOGRAD_KEY not in self.attrs and not isbox(flat := self.data.flat[0]):
        return None
    elif isbox(flat):
        return anp.array(self.values.tolist())
    else:
        return self.attrs[AUTOGRAD_KEY]
```
Compared to
```python
@property
def tracers(self) -> Box:
    if self.data.size == 0:
        return None
    else:
        flat = self.data.flat[0]
        if AUTOGRAD_KEY not in self.attrs and not isbox(flat):
            return None
        elif isbox(flat):
            return anp.array(self.values.tolist())
        else:
            return self.attrs[AUTOGRAD_KEY]
```
Note that we need to add this level of nesting since `self.data.flat[0]` might error when `self.data.size == 0`.

Reply:
You don't need `else` after `return`?

Reply:
oh man 😂  yeah